### DISCUSSION TITLE
Virtual version name convention for package pre-release version

### DISCUSSION THREAD

**Discussion by cyfdecyf** [xmake-io/xmake-repo#1112] [General]
As discussed in #1102, we need to use pre-release commit for some packages.

**Do we have convention on version names for pre-release commit?**

Let's use`rapidjson` for example, the latest release is in 2016 but has many more commits after that release. I'm currently using a version name `v1.10.0-arrow`. This won't affect the default version user will get, but this name is for specific use of apache arrow.

I suggest appending a short commit id as suffix for virtual version. For example:

```lua
add_versions("v1.1.0-3c04e1", "3c04e1ed8dda45b9dc58a64e3cf1275df2d2d7482fd6fea9949636a574558310")
```

For packages that has never made a release, seems like we are currently using the date for the commit.

**Comment by waruqi** [comment]
对于没有发布过得包，默认约定按日期来，用户即使不知道 commit ，也能大概知道版本的大概日期，也方便根据日期时间做版本选择, ">= 2021.01"。。

如果只是对已有 version list ，加额外的 commit 的特殊修订版本，最好还是按更加可读的 name 来标识，修复的特性，例如 v1.1.0-feature

让用户看一眼版本就能知道，这个特殊版本，带了什么修复。。没必要加 commit 进去，不可读，选择版本时候，用户还得去反查

可以参考 imgui 的特殊版本命名。

https://github.com/xmake-io/xmake-repo/blob/4367726d063ce6a933a4af94ce7bab343f37e55a/packages/i/imgui/xmake.lua#L9-L16

**Reply by cyfdecyf** [reply]
那像 `v1.10.0-arrow` 这样的名字可以接受么？说明是 arrow 这个包使用的版本。

**Reply by waruqi** [reply]
> 那像 `v1.10.0-arrow` 这样的名字可以接受么？说明是 arrow 这个包使用的版本。

如果这是一个特定于对 arrow 的版本，可以这么命名，如果是通用的 prerelease 版本，也许其他库使用，可以用 `v1.10.0-pre` 啥的

**Comment by leodido** [comment]
Wondering whether we could use this also for `lz4` (eg., `v.1.9.3-arrow`). In fact, Apache Arrow uses commit 8f61d8eb7c6979769a484cde8df61ff7c4c77765 (see https://github.com/apache/arrow/blob/ccaef092c297b8af5d375bc542d908cda8fd415e/cpp/thirdparty/versions.txt#L61)

**Reply by waruqi** [reply]
maybe we should find a better solution. like this

```lua
package("arrow")
    add_deps("lz4 8f61d8eb7c6979769a484cde8df61ff7c4c77765")
    add_deps("rapidjson 8f61d8eb7c6979769a484cde8df61ff7c4c77765")
    -- ...
end
```

but It has been not supported yet.

**Reply by waruqi** [reply]
I have supported it on dev branch.

**Reply by cyfdecyf** [reply]
This would be very useful for testing with unreleased commit. I tried to use commit id as version in `add_requireconfs` as in this commit https://github.com/cyfdecyf/xrepo-apache-arrow/commit/ffa59b9d3178ac910dbdce8f52cec3ffd2159f3d, but got following error:

```
package(rapidjson): version(1a803826f1197b5e30703afe4b9c0e7dd48074f5) not found!
```

Changing `arrow/xmake.lua` in xmake-repo got the same error result. 

Is this expected?

**Reply by waruqi** [reply]
try it again on dev

**Reply by cyfdecyf** [reply]
Works now. Thanks!
